src/vos/vos_obj.c

```diff
@@ -74,17 +74,22 @@ key_punch(struct vos_object *obj, daos_epoch_t epoch, uint32_t pm_ver,
 				    flags);
 		if (rc != 0)
 			D_GOTO(out, rc);
-
 	} else {
 		daos_handle_t		 toh;
 		int			 i;
 
-		rc = key_tree_prepare(obj, obj->obj_toh, VOS_BTR_DKEY,
-				      dkey, SUBTR_CREATE, DAOS_INTENT_PUNCH,
-				      &krec, &toh);
+		/* For non-rebuild case, if the key does not exist,
+		 * don't create it.
+		 */
```

Comment:
I don't quite see why we want to introduce this complexity (it looks actually not quite right to me). DAOS native key punch is essentially a versioned update, and I think we'd keep this semantics for correctness and simplicity.

Reply:
What is the semantics for negative object/key punch? If we fetch the object/key with the epoch that we used for the negative object/key punch, what is the expected result? Any difference from the case of without such object/key? I do not think it will cause any correctness issue.

As for complicity, I am not sure what you mean, as my understand, creating the object/key for negative punch is more complex than the patch does, because it touches unnecessary incarnation log and DTX logic. Skip such negative punch will simplify the work for both incarnation log and DTX, and potential rebuild and vos aggregation.

Reply:
> What is the semantics for negative object/key punch?

I think we should just simply log the punch, no? If it's conditional update, we may return error in this case.

> Any difference from the case of without such object/key? I do not think it will cause any correctness issue.

Assume two concurrent operations: update & punch to same key, and we don't guarantee the two requests are process in epoch order on server, the punch could be lost with your patch, right?

Reply:
> > What is the semantics for negative object/key punch?
> 
> I think we should just simply log the punch, no? If it's conditional update, we may return error in this case.

The direct benefit is to simplify the incarnation log for punch, the in-direct benefit is also for rebuild and vos aggregation.

> 
> > Any difference from the case of without such object/key? I do not think it will cause any correctness issue.
> 
> Assume two concurrent operations: update & punch to same key, and we don't guarantee the two requests are processed in epoch order on server, the punch could be lost with your patch, right?

Since we do not care about the executed order, then what is the expected result? What's wrong if we lost the punch (if it is executed firstly) with my patch?